### PR TITLE
revamp withdraw manager

### LINKED ISSUE #255: Double claim in withdraw manager
We didn't mark the token is claimed in ClaimTokens 🙈.

Ideally, to save storage usage and gas, we should introduce a bitmap like:
 
https://github.com/Uniswap/merkle-distributor/blob/master/contracts/MerkleDistributor.sol#L13

### PR DESCRIPTION
**Pull request by ChihChengLiang**
Fix #255

### CODE REVIEW

File: test/MassMigration.test.ts [hunk 1]
```diff
@@ -150,4 +130,100 @@ describe("Mass Migrations", async function() {
...
+
+        await expect(
+            withdrawManager.ProcessWithdrawCommitment(batchId, batch.proof(0))
+        ).revertedWith("Vault: Batch shoould be finalised");
+
+        await mineBlocks(ethers.provider, TESTING_PARAMS.TIME_TO_FINALISE);
+
+        // We cheat here a little bit by sending token to the vault manually.
+        // Ideally the tokens of the vault should come from the deposits
+        await testToken.transfer(vault.address, tx.amount);
```
Comment (ChihChengLiang) [inline test/MassMigration.test.ts:177]: Haven't had a way to finalise deposit well with current utils, so doing a manual transfer here.

File: ts/constants.ts [hunk 2]
```diff
@@ -3,7 +3,8 @@ import { DeploymentParameters } from "./interfaces";
 export const TESTING_PARAMS: DeploymentParameters = {
     MAX_DEPTH: 4,
     MAX_DEPOSIT_SUBTREE_DEPTH: 1,
-    STAKE_AMOUNT: "0.1"
+    STAKE_AMOUNT: "0.1",
+    TIME_TO_FINALISE: 5
```
Comment (ChihChengLiang) [inline ts/constants.ts:7]: We can customize TIME_TO_FINALISE in tests now.

File: ts/deploy.ts [hunk 3]
```diff
@@ -170,6 +165,28 @@ export async function deployAll(
...
+    const spokeRegistry = await new SpokeRegistryFactory(signer).deploy();
+    await waitAndRegister(
+        spokeRegistry,
+        "spokeRegistry",
+        verbose,
+        nameRegistry,
+        await paramManager.SPOKE_REGISTRY()
+    );
+
+    const vault = await new VaultFactory(allLinkRefs, signer).deploy(
```
Comment (ChihChengLiang) [inline ts/deploy.ts:180]: Gotchas here: 
Vault depends on spokeRegistry and Deposit Manager depends on Vault. If we deploy with the wrong order, the transactions fail without giving any reason.
 Created an issue in #270
A workaround is always requiring non-Zero contract address in the constructor.

File: ts/utils.ts [hunk 4]
```diff
@@ -40,6 +40,15 @@ export function randomLeaves(num: number): string[] {
     return leaves;
 }
 
+export async function mineBlocks(
```
Comment (ChihChengLiang) [inline ts/utils.ts:43]: This util allows us to time travel n blocks

**Comment by ChihChengLiang** [comment]
Ah ah, running into a dependency hell ...
Vault depends on rollup, rollup depends on deposit manager, and deposit manager depends on vault. 😱

File: contracts/Vault.sol [hunk 5]
```diff
@@ -30,6 +27,15 @@ contract Vault {
...
 
+    /**
+    @dev We assume Vault is deployed before Rollup
+     */
+    function setRollupAddress() external {
+        rollup = Rollup(
+            nameRegistry.getContractDetails(ParamManager.ROLLUP_CORE())
+        );
+    }
+
```
Comment (ChihChengLiang) [inline contracts/Vault.sol:38]: Ugly, ... but let's figure it out later.

**Review by vaibhavchellani — APPROVED** [review]
Looks good to merge to me!

**Pull request merged**